Fix configure/make errors on cygwin

tobil4sk opened the pull request:
- Cygwin struggles to handle carriage returns in `ocamlc -v` output
- Some $(EXE) extensions were missing
- `obj` and `lib` extensions were being configured incorrectly

chetmurthy commented:
Hi, I haven't had a chance to check that this works OK, but I see it's pretty simple, so will do so soon.  And it'll be part of the next bugfix release.  Thank you for this!

tobil4sk commented:
Hi Chet, thanks! I've added a few more patches needed to fix some further issues for a successful build on cygwin. There is also this related issue: https://github.com/camlp5/camlp5-buildscripts/pull/1.

As an additional note, when compiling I had to make sure that the "perl.exe" available in PATH is the cygwin one rather than the native windows version I also have. This fixes the perl script (so I have reverted the unneeded change in 2f69341cebbda4c083f222eb9cc3cd74e862343d). Using the wrong perl build also completely breaks the output of [strip-parser.pl](https://github.com/camlp5/camlp5/blob/master/compile/strip-parser.pl#L6) script which results in a broken `pa_o_fast.ml`:

```
- ocamlfind ocamlopt -package compiler-libs,camlp-streams,rresult,bos,re -g -I ../main -I ../boot -I ../lib -c -impl pa_o_fast.ppo
- File "pa_o_fast.ml", line 1720, characters 33-48:
- 1720 |                        [ [: op = operator_rparen; loc = P.gloc bp ! :] ->
-                                         ^^^^^^^^^^^^^^^
- Error: This expression has type "string Grammar.Entry.e"
-        This is not a function; it cannot be applied.
- make[2]: *** [../config/Makefile:32: pa_o_fast.cmx] Error 2
```

So together with these patches and ensuring that the correct perl build is available in PATH, things seem to be compiling correctly on cygwin now!

tobil4sk commented:
Hi @chetmurthy, any chance it may be possible to look into getting this merged?

chetmurthy commented:
Oh, looks like I merged a previous version of this patch back in Oct 2024.  I'll look at merging this now.  It's getting to be time for a new release anyway, thank you for mentioning this!

chetmurthy commented:
I want to thank you again for prodding me to get this Windows compat code merged and released.

a2line commented:
Hello. I'm very happy to see work on Windows side. I tried to pin the current camlp5 master and run it against Geneweb CI under Windows only, and for multiple environnements. I tried ocaml-setup v2 and v3, knowing that v3 was failing only on building camlp5.8.03.00 (I retested that two weeks ago):

It's failing on v2 because repository is not up to date and doesn't have camlp5-buildscripts, this is normal and good, it allows to build on Windows with old 8.00.03 campl5.
``` bash
[camlp5.8.00.03] synchronised from git+https://github.com/camlp5/camlp5.git#master
[WARNING] Failed checks on camlp5 package definition from source at git+https://github.com/camlp5/camlp5.git#master:
    error  3: File format error in 'url.checksum' at line 67, column 4: while expecting checksum: Invalid_argument("OpamHash.of_string")
camlp5 is now pinned to git+https://github.com/camlp5/camlp5.git#master (version 8.03.02)

[NOTE] Pinning command successful, but your installed packages may be out of sync.
The following dependencies couldn't be met:
  - camlp5 -> camlp5-buildscripts
      unknown package
```

I know that ocaml-setup@v3 is also failing with Ocaml 4.08, so I was hoping on last two builds on v3 (ocaml4.14.2/5.3.0) to work, I see a [small error building](https://github.com/a2line/geneweb/actions/runs/13932380641/job/38992452744#step:8:1):
``` bash
Run opam pin add -y camlp5 https://github.com/camlp5/camlp5.git#master
[camlp5.8.03.01] synchronised (git+https://github.com/camlp5/camlp5.git#master)
[WARNING] Failed checks on camlp5 package definition from source at git+https://github.com/camlp5/camlp5.git#master:
    error  3: File format error in 'url.checksum' at line 67, column 4: while expecting checksum: Invalid_argument("OpamHash.of_string")
camlp5 is now pinned to git+https://github.com/camlp5/camlp5.git#master (version 8.03.02)

[ERROR] In the opam file for camlp5.8.03.02:
          - At D:\a\geneweb\geneweb\_opam\.opam-switch\sources\camlp5\opam:67:4-67:13::
            while expecting checksum: Invalid_argument("OpamHash.of_string")
        'url.checksum' has been ignored.

```

It seems Camlp5 get installed. We have a build error later building Geneweb, build error we don't have on 8.03.00 release nor on 8.00.03 used for Windows on current passing CI.

``` bash
File "bin/ged2gwb/dune", line 2, characters 8-15:
2 |   (name ged2gwb)
            ^^^^^^^
Error: File unavailable: "D:agenewebgeneweb_opamlib/camlp5\\camlp5.cmxa"
File "bin/ged2gwb/dune", line 2, characters 8-15:
2 |   (name ged2gwb)
            ^^^^^^^
Error: File unavailable: "D:agenewebgeneweb_opamlib/camlp5\\odyl.cmxa"
File "bin/ged2gwb/dune", line 5, characters 27-40:
5 |   (preprocess (action (run not-ocamlfind preprocess -package camlp5.extend,camlp5.quotations,camlp5.pr_o -syntax camlp5o %{input-file})))
                               ^^^^^^^^^^^^^
Error: Program not-ocamlfind not found in the tree or in PATH
 (context: default)
```
I'm trying a second time to force installation of not-ocamlfind but the first path we see is very bad.

chetmurthy commented:
@a2line I don't quite understand what you've posted.  `camlp5` does not use `not-ocamlfind` as part of its build or tests.  So this must be used by geneweb ?  Can you clarify what builds/installs, and what does not?

Also, the "OpamHash" warning you're getting is normal when you pin to a GIT repo: it contains the opam file, and that opam file has the stanza for the URL and hash.  I do that so I can mechanically fill it in during release (with a program).  Here it is below
```
url {
  src: ""
  checksum: [
    "sha512="
  ]
}
```

tobil4sk commented:
> Hello, may I ask that you try out the current master on Windows and see if it works ?

I got caught out again by not having cygwin's perl available in PATH, and I also got the error about the checksum, but aside from that it installs! Thanks for merging and submitting so swiftly!

> may I ask if you would be willing to do the same with pa_ppx ?

Sure, I will have a look into it! I'm running into some initial problems, I will open an issue over there.

chetmurthy commented:
> ```shell
> "D:agenewebgeneweb_opamlib/camlp5\\camlp5.cmxa"
> ```

Is this what you mean by "the first path we see is very bad" ?  I don't know where that path comes from: whether in Camlp5 or in geneweb.  But I would be happy to help you debug.  Since I don't have Windows (or Windows knowledge/experience) I'll have to be helping you debug, though.

a2line commented:
Sorry, only first part about checksum error on campl5 build was relevant for you, and as you said once released that error message will probably disappear.

Camlp5 is now building ocaml 5.3.0 (and 4.14.2 too, where it shouldn't I think ?!) with ocaml-setup@v3 for Actions CI (note that v3 doesn't support ocaml4.08 at the moment).

At end of https://github.com/geneweb/geneweb/pull/1806#discussion_r1630425218 I prooved not-ocamlfind were you helped us last year I showed that not-ocamlfind is instaslled doing `opam install camlp5`, I'm quite was quite sure of that, but now I'm not anymore…

chetmurthy commented:
BTW, I  can read French, so if you prefer, please go ahead and write in French.

chetmurthy commented:
> Camlp5 is now building for Windows with Ocaml 5.3.0 (and 4.14.2 too, where it shouldn't I think ?!)

I am not surprised that it builds for many different versions of OCaml: @tobil4sk 's patch was quite general!  *grin*

tobil4sk commented:
The path issue looks quite similar to the issue I had to solve by adding quotes around the paths before passing them through the cygwin shell.

Regarding not-ocamlfind, I noticed that it installs the binary without an .exe suffix, it could be why dune is failing to find it.

> Sure, I will have a look into it! I'm running into some initial problems, I will open an issue over there.

By the way, it seems the camlp5/pa_ppx fork does not have issues enabled, where would be the best place to open an issue?

chetmurthy commented:
I'll look at not-ocamlfind for the .exe thing.

I noticed that pa_ppx did not have issues enabled, and have enabled it now.  I will go thru all my repos and ensure that they all have issues enabled (I thought they were by default!)

tobil4sk commented:
> I'm assuming you mean "in the Camlp5 repository" ?

Indeed, if support for ocaml 4.08 is not a concern, then it is possible to simply use something like what camlp5 already does:

https://github.com/camlp5/camlp5/blob/3686462c9abe18c938f1bf535a7d181b4ede8480/configure#L197

Previously it parsed the full output of `ocamlc -config`, but that didn't work well with carriage returns on windows.

chetmurthy commented:
> Indeed, if support for ocaml 4.08 is not a concern, then it is possible to simply use something like what camlp5 already does

I looked at `ocamlfind` (which not-ocamlfind includes and builds in order to have certain libraries available) and its "configure" process deduces the same things.  I think I'll just include their Makefile.config -- I depend on it being correct anyway.

chetmurthy commented:
@a2line If I prepare a new version of not-ocamlfind with the proper file-extension (.exe), could I ask you to try it out (on windows) ?

Also, I'd like to ensure that your software is working with these new packages.

a2line commented:
Yes, I would of course test anything that could help!

chetmurthy commented:
OK, so we can communicate about this, could you open an issue over at not-ocamlfind?

https://github.com/chetmurthy/not-ocamlfind/issues

chetmurthy commented:
@tobil4sk  since you obviously know about Windows<->UNIX compatibility, could I impose upon you to ask some more questions ?

the first one would be: In the camlp5 Makefiles, I don't take any care to use "\\" instead of "/" -- more precisely, I haven't got a "DIRSEP" variable or anything like that, and yet it works?  I think this means that most of the code being executed is indifferent to the directory-separator (or switches them automatically?)

tobil4sk commented:
Cygwin is able to handle unix like paths normally, the problem is when paths come from elsewhere, for example from ocaml compiler variables which are in native windows format instead of unix format. Cygwin's shell strips out any `\` it finds because it reads them as escape characters, which is why they disappear after being processed. You can avoid this issue by putting quotes around the paths that may be in native windows format, which prevents cygwin's shell from removing the `\`.

chetmurthy commented:
Oh I see!  OK.  Two follow-ups:

(1) So there are places where I'm generating paths like this:
```
	OVERSION=$(OVERSION) VERSION=$(VERSION) CAMLP5DIR="$(LIBDIR)/$(CAMLP5N)" ./META.pl > META
```
(that variable CAMLP5DIR) and LIBDIR is generated by "configure" (so I guess we can assume it contains "\\" and not "/" ?)  should I be using a DIRSEP or should I let the perl-script do the fixup?

(2) and I see in camlp5's configure:
```
if test "$libdir" != ""; then
    LIBDIR=$libdir
elif test "$prefix" != ""; then
    LIBDIR='$(PREFIX)/lib/ocaml'
else
    LIBDIR=$ocamlc_where
fi
```
and I wonder: let's assume PREFIX is in windows-format; what happens when "/lib/ocaml" gets appended ?  Do I need to fixup all these paths also? [grumble grumble I wanna kick Bill Gates in the balls so-so-so-so hard!]

tobil4sk commented:
> (that variable CAMLP5DIR) and LIBDIR is generated by "configure" (so I guess we can assume it contains "\" and not "/" ?) should I be using a DIRSEP or should I let the perl-script do the fixup?

It looks like in the specific case of the META, the path needs to be treated with care and all instances of `\` should be replaced by either `/` or `\\`.

> and I wonder: let's assume PREFIX is in windows-format; what happens when "/lib/ocaml" gets appended ? Do I need to fixup all these paths also? [grumble grumble I wanna kick Bill Gates in the balls so-so-so-so hard!]

In general I think this should not be an issue, it's just in specific cases when printing the path to META where it may turn out to be a problem.

chetmurthy commented:
Now I'm wondering about what will happen if I "fix" it for recent OCaml versions (e.g. 5.3.0).  There, I need to use "\\".  But apparently it was working somewhat correctly for older OCaml versions, so I wonder if (e.g.) for OCaml 4.14.1 that fix will break things, b/c instead of
```
C:\a\b\c/camlp5
```
I'd get
```
C:/a/b/c\camlp5
```
And again, I'd get findlib breakage.

chetmurthy commented:
@tobil4sk So I looked at the Makefile.cnf (generated by the Camlp5 configure script) files that @A2 sent me.  And it seems like:

(1) `ocamlc -where` is generating paths in Cygwin format, e.g.
```
OVERSION=5.3.0
OLIBDIR=D:/a/geneweb/geneweb/_opam/lib/ocaml
```
and
```
OVERSION=4.14.0
OLIBDIR=C:/cygwin/home/A2/.opam/4.14.2+mingw64/lib/ocaml
```
those files came from Ocaml 5.3.0 and 4.14.0, and in both cases OLIBDIR was from `ocamlc -where`.

(2) But PREFIX is completely different:
```
OVERSION=4.14.0
PREFIX=C:/cygwin/home/A2/.opam/4.14.2+mingw64
```
and
```
OVERSION=5.3.0
PREFIX=D:\a\geneweb\geneweb\_opam
```

(3) Hmmm .... I wonder where PREFIX comes from?  I -think- it's coming from opam, which supplies prefix, libdir, etc, right?  So I wonder if the opam version changed between when @A2 did her 4.14.0 build, and her 5.3.0 build ?  Was it a different OPAM setup?  Looking at the CI builds I see (for the 4.14.2 mingw build):
```
C:\Windows\system32\cmd.exe /D /S /C "D:\cygwin\wrapperbin\opam.cmd --version"
2.0.10
```
and for the 5.3.0 build:
```
C:\hostedtoolcache\windows\opam\2.3.0\x86_64\opam.exe --version
2.3.0
```
[I confess that I'm having trouble making sure I go to the right build each time, because there are at least two 5.3.0 builds]

So it looks like the build that had Cygwin paths for PREFIX,etc was opam 2.0, and the build that had Windows-style paths, was opam 2.3.  I cannot see the "configure" line that was passed to camlp5, but if I could, maybe that would shed some light on why the older Camlp5 worked, and the newer one does not.

@A2, do you have any comments on this?

chetmurthy commented:
With @a2line 's help, we were able to figure out what was wrong and fix it.  Huzzah!

tobil4sk commented:
> I -think- it's coming from opam, which supplies prefix, libdir, etc, right? 

Yes, it is an opam variable: https://github.com/camlp5/camlp5/blob/1716052af6b611131456358c758c0294e3a6b4d8/opam#L46

> Was it a different OPAM setup

setup-ocaml@v2 used to use a cygwin installed opam, while setup-ocaml@v3 uses a natively windows opam, so this is likely to be the case.

> With a2line 's help, we were able to figure out what was wrong and fix it. Huzzah!

Great! Some of those patches will also help with pa_ppx. I've opened an issue over in that repo for the next problem I ran into after that: https://github.com/camlp5/pa_ppx/issues/2

chetmurthy commented:
Hello, could you email me at the maintainer address of this package?  I would like to send you a message privately.
